pkg/internal/cnsoperator/cnsfilevolumeclient/cnsfilevolumeclient.go

```diff
@@ -0,0 +1,251 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cnsfilevolumeclient
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/logger"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/internal/cnsoperator/cnsfilevolumeclient/v1alpha1"
+	k8s "sigs.k8s.io/vsphere-csi-driver/pkg/kubernetes"
+)
+
+// FileVolumeClient exposes an interface to support
+// configuration of CNS file volume ACL's.
+type FileVolumeClient interface {
+	// AddClientVMToIPList adds the input clientVMName to the list of
+	// clientVMNames that expose the same external clientVMIP for a
+	// given file volume. fileVolumeName is used to uniquely
+	// identify CnsFileVolumeClient instances.
+	AddClientVMToIPList(ctx context.Context, fileVolumeName, clientVMName, clientVMIP string) error
+	// RemoveClientVMFromIPList removes the input clientVMName from
+	// the list of clientVMNames that expose the same external
+	// clientVMIP for a given file volume. fileVolumeName is used
+	// to uniquely identify CnsFileVolumeClient instances.
+	RemoveClientVMFromIPList(ctx context.Context, fileVolumeName, clientVMName, clientVMIP string) error
+}
+
+// fileVolumeClient maintains a client to the API
+// server for operations on CnsFileVolumeClient instance.
+// It also contains a per instance lock to handle
+// concurrent operations.
+type fileVolumeClient struct {
+	client client.Client
+	// Per volume lock for concurrent access to CnsFileVolumeClient instances.
+	// Keys are strings representing volume handles (or SV-PVC names).
+	// Values are individual sync.Mutex locks that need to be held
+	// to make updates to the CnsFileVolumeClient instance on the API server.
+	volumeLock *sync.Map
+}
+
+var (
+	fileVolumeClientInstanceLock sync.Mutex
+	fileVolumeClientInstance     *fileVolumeClient
+)
+
+// GetFileVolumeClientInstance returns a singleton of type FileVolumeClient.
+// Initializes the singleton if not already initialized.
+func GetFileVolumeClientInstance(ctx context.Context) (FileVolumeClient, error) {
+	fileVolumeClientInstanceLock.Lock()
+	defer fileVolumeClientInstanceLock.Unlock()
+	if fileVolumeClientInstance == nil {
+		log := logger.GetLogger(ctx)
+		config, err := k8s.GetKubeConfig(ctx)
+		if err != nil {
+			log.Errorf("failed to get kubeconfig. Err: %v", err)
+			return nil, err
+		}
+		k8sclient, err := k8s.NewClientForGroup(ctx, config, v1alpha1.GroupVersion.Group)
+		if err != nil {
+			log.Errorf("failed to create k8s client. Err: %v", err)
+			return nil, err
+		}
+		informer, err := k8s.GetDynamicInformer(ctx, v1alpha1.GroupVersion.Group, v1alpha1.GroupVersion.Version, "cnsfilevolumeclients")
```

Comment:
Why do we need the informer now given that we are not using the cache?

Reply:
Removed